### PR TITLE
feat: add win condition

### PR DESCRIPTION
**Pull request by JoenvandeVorle**
## Description

Adds a win screen, which will be shown if you enter the win area while carrying the Eternal Elixir

## Testability

You can go to the `WinCondition` scene to easily test the win screen. If you enter the `GridSystem` scene and generate/complete a level, you should be able to do the same.

## Checklist
- [x] Set the proper pull request name
- [x] Merged main into your branch
- [x] Added a scene to the game for your changes

### CODE REVIEW

File: aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs [hunk 1]
```diff
@@ -119,15 +105,26 @@ private void MakeScene()
...
+            if (_useSeed)
+            {
+                _random = new Random(_seed);
+            }
+            else
+            {
+                Random tempRandom = new();
+                _seed = tempRandom.Next();
+                _random = new Random(_seed);
+            }
```
Comment (joachimdekker) [inline aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs:117]: Use conditional inverting to simplify this code.
```suggestion
            if (!_useSeed)
            {
                Random tempRandom = new();
                _seed = tempRandom.Next();
            }
            _random = new Random(_seed);
            
```

**Review by joachimdekker — CHANGES_REQUESTED** [review]
Looks good. See my comment. Have to test it still.

**Review by joachimdekker — APPROVED** [review]
Everything works as intended. Good job both.

**Review by QuakeEye — CHANGES_REQUESTED** [review]
It is currently possible to softlock the game by using (apparently) the bottle that you have to return (apparently).

File: aplib.net-demo/Assets/Scenes/GridSystem.unity [hunk 2]
```diff
@@ -611,6 +611,72 @@ Transform:
...
+  m_ExcludeLayers:
+    serializedVersion: 2
+    m_Bits: 0
+  m_LayerOverridePriority: 0
+  m_IsTrigger: 1
+  m_ProvidesContacts: 0
+  m_Enabled: 1
+  serializedVersion: 3
+  m_Size: {x: 1, y: 1, z: 1}
+  m_Center: {x: 0, y: 0, z: 0}
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scenes/GridSystem.unity:679]: The Win Area object does not seem to work, I collected the end item and went into this area, and nothing happened.

File: aplib.net-demo/Assets/Prefabs/CanvasManager WinScreen.prefab
Comment (QuakeEye) [inline aplib.net-demo/Assets/Prefabs/CanvasManager WinScreen.prefab:1]: There are multiple canvasses active, causing z-clipping.

File: aplib.net-demo/Assets/Scenes/GridSystem.unity
[the remark this answers is not in the document]
Comment (Tboefijn) [reply aplib.net-demo/Assets/Scenes/GridSystem.unity:1]: Heb je de seed van de keer dat de end item niet is ingespawned?

File: aplib.net-demo/Assets/Scripts/MainMenu/CanvasManager.cs [hunk 3]
```diff
@@ -69,12 +74,19 @@ private void Awake()
     private void Start()
     {
         _currentSceneName = SceneManager.GetActiveScene().name;
+        WinArea winArea = FindObjectOfType<WinArea>();
+        if (winArea != null)
+        {
+            winArea.OnWin += ShowWinScreen;
+
+        }
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/MainMenu/CanvasManager.cs:82]:
```suggestion
        if (winArea != null)
            winArea.OnWin += ShowWinScreen;
```

File: aplib.net-demo/Assets/Scripts/WFC/GameObjectPlacer.cs [hunk 4]
```diff
@@ -112,8 +138,12 @@ public void Initialize()
         /// </summary>
         /// <param name="cell">The cell to spawn the end item in.</param>
         /// <param name="parent">The parent of the end item.</param>
-        public void PlaceEndItem(Cell cell, Transform parent) =>
+        public void PlaceEndItem(Cell cell, Transform parent)
+        {
             Instantiate(_endItemPrefab, CenterOfCell(cell) + _floorOffset, Quaternion.identity, parent);
+            // Change the cell color to end room color
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/WFC/GameObjectPlacer.cs:144]:
```suggestion
            // Change the cell color to end room color.
```

File: aplib.net-demo/Assets/Scripts/WFC/GameObjectPlacer.cs [hunk 5]
```diff
@@ -237,5 +267,41 @@ public void PlaceTile(int x, int z, Tile tile, Transform parent)
...
+
+            GameObject winPoint = GameObject.FindWithTag("Win");
+
+            if (winPoint == null) return;
+
+            winPoint.transform.position = spawningPoint;
+            Area winArea = winPoint.GetComponent<Area>();
+            winArea.Bounds = new Bounds(spawningPoint, winArea.Bounds.extents);
+
+            // Set the colors of the start and end rooms
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/WFC/GameObjectPlacer.cs:303]:
```suggestion
            // Set the colors of the start and end rooms.
```

**Review by QuakeEye — CHANGES_REQUESTED** [review]

**Comment by JoenvandeVorle** [comment]
> It is currently possible to softlock the game by using (apparently) the bottle that you have to return (apparently).

as intended

File: aplib.net-demo/Assets/Prefabs/CanvasManager WinScreen.prefab
Comment (JoenvandeVorle) [reply aplib.net-demo/Assets/Prefabs/CanvasManager WinScreen.prefab:1]: Which scene, which canvases? Can you take a screenshot?

**Comment by QuakeEye** [comment]
> > It is currently possible to softlock the game by using (apparently) the bottle that you have to return (apparently).
> 
> as intended

There should be some way to cancel or restart, softlocking should never be possible.

File: aplib.net-demo/Assets/Scenes/GridSystem.unity (hunk 2 above)
Comment (JoenvandeVorle) [reply aplib.net-demo/Assets/Scenes/GridSystem.unity:861]: Forgot to add the canvasmanager in this scene, should work now

File: aplib.net-demo/Assets/Prefabs/CanvasManager WinScreen.prefab
Comment (QuakeEye) [reply aplib.net-demo/Assets/Prefabs/CanvasManager WinScreen.prefab:1]: The file I made the comment on?

File: aplib.net-demo/Assets/Scenes/GridSystem.unity
[the remark this answers is not in the document]
Comment (QuakeEye) [reply aplib.net-demo/Assets/Scenes/GridSystem.unity:1]: It didn't work about half of the time, now that the level size has been increased it doesn't seem to be happening anymore, but the chance might be non-zero?

**Review by joachimdekker — APPROVED** [review]
All works as intended. With Enemies, I foresee a _fun_ game!

File: aplib.net-demo/Assets/Scenes/GridSystem.unity
[the remark this answers is not in the document]
Comment (QuakeEye) [reply aplib.net-demo/Assets/Scenes/GridSystem.unity:1]: 1238555218

Comment (QuakeEye) [inline aplib.net-demo/Assets/Scenes/GridSystem.unity:1]: After winning, it is possible to go to the menu, then out of the menu, you are now able to move around again while already having won and you can also use the end item (when death screen is implemented the death screen may then be visible while the win screen is also visible?)

**Review by QuakeEye — CHANGES_REQUESTED** [review]
Once the existing comments have been fixed together with the index error I can approve this

**Review by joachimdekker — APPROVED** [review]
Children Cheering

**Review by SSelinn — APPROVED** [review]
Looking good

**Comment by joachimdekker** [comment]
afbeelding
I get this error when running the grid system. It seems like a one-off error, but is this a known issue? Furthermore, it seems like item spawning does not work.

**Review by joachimdekker — APPROVED** [review]
DALL·E 2024-05-29 20 01 16 - A vibrant, cheerful image with the text 'There are the items! Now I can approve!' in bold, colorful letters  The background should feature an abstract (Medium)

**Pull request merged**